Zephyr toolchain / linking improvements

tejlmand opened the pull request:
This is a series of linking improvements in Zephyr.

First part is moving linking of C and runtime libraries out away from `zephyr_link_libraries()`.
Using `zephyr_link_libraries()` is vulnerable to link dependency and link ordering, which has been discussed multiple times in the past.
It has also resulted in CMake construct like this: `zephyr_link_libraries(-lc)` / `zephyr_link_libraries(-lgcc)` and thus making it harder to support non-gcc toolchains / libraries. This again has resulted in constructs like `if(${ZEPHYR_TOOLCHAIN_VARIANT}) ...)`. All of this should be handled by Zephyr's toolchain infrastructure and normal CMake library creation should not need to know anything about C or runtime libraries, nor the toolchain in use.

By moving C / runtime library linking into the toolchain infrastructure and specify C / runtime libraries using CMake linker rules instead of `zephyr_link_libraries()` then we avoid CMake reordering the libraries, and CMake deduplication, and we ensure they are always linked at correct location.
To still allow Zephyr CMake code full flexibility and the use of Kconfig to select the libraries to use, then standard library selection has been moved to CMake linker properties, which provides a given set of defaults depending on the toolchain in use.
But the default can later be adjusted, based on other criteria.

This set of improvements allows for better LLVM support, especially when using LLVM with picolibc built-in support, but also when building picolibc as a module.

With this PR it is now possible to use LLVM for Arm for cross-compilation using both built-in Picolibc support as well as picolibc as module. LLVM for Arm with newlib is also supported.

LLVM for Arm:
https://github.com/ARM-software/LLVM-embedded-toolchain-for-Arm/releases/

In environment, set:
```
$ export ZEPHYR_TOOLCHAIN_VARIANT=llvm
$ export LLVM_TOOLCHAIN_PATH=/<path-to-LLVM-for_Arm_installation>/
# Example: export LLVM_TOOLCHAIN_PATH= /opt/LLVM-ET-Arm-18.1.3-Linux-x86_64/
```

To build Zephyr Hello world for an Arm target, such as nrf52840dk, using build-in picolibc support and lld with compiler rt, use:
```
$ west build -p -b nrf52840dk/nrf52840 samples/hello_world -- -DCONFIG_BUILD_NO_GAP_FILL=y -DCONFIG_COMPILER_RT_RTLIB=y -DCONFIG_LLVM_USE_LLD=y
```

Using picolibc as module:
```
$ west build -p -b nrf52840dk/nrf52840 samples/hello_world -- -DCONFIG_BUILD_NO_GAP_FILL=y -DCONFIG_COMPILER_RT_RTLIB=y -DCONFIG_LLVM_USE_LLD=y -DCONFIG_PICOLIBC_USE_MODULE=y
```

Using newlib as pre-built (remember to also install `LLVM-ET-Arm-newlib-overlay-18.1.3.tar.xz` ):
```
$ west build -p -b nrf52840dk/nrf52840 samples/hello_world -- -DCONFIG_BUILD_NO_GAP_FILL=y -DCONFIG_COMPILER_RT_RTLIB=y -DCONFIG_LLVM_USE_LLD=y -DCONFIG_NEWLIB_LIBC=y
```

As part of the LLVM improvement, then support for LLVM with C++ has also been improved, which means it's now possible to build the simply C++ hello world:
```
$ west build -p -b nrf52840dk/nrf52840 samples/cpp/hello_world -- -DCONFIG_BUILD_NO_GAP_FILL=y -DCONFIG_COMPILER_RT_RTLIB=y -DCONFIG_LLVM_USE_LLD=y
...
[137/137] Linking CXX executable zephyr/zephyr.elf
ld.lld: warning: /opt/LLVM-ET-Arm-18.1.3-Linux-x86_64/bin/../lib/clang-runtimes/arm-none-eabi/armv7em_soft_nofp_exn_rtti/lib/libc++abi.a(stdlib_new_delete.cpp.obj):(__lcxx_override) is being placed in '__lcxx_override'
ld.lld: warning: <internal>:(.eh_frame) is being placed in '.eh_frame'
ld.lld: warning: ignoring memory region assignment for non-allocatable section '.last_section'
Memory region         Used Size  Region Size  %age Used
           FLASH:      173484 B         1 MB     16.54%
             RAM:        8448 B       256 KB      3.22%
        IDT_LIST:          0 GB        32 KB      0.00%
Generating files from /projects/github/ncs/zephyr/build/zephyr/zephyr.elf for board: nrf52840dk
```
although some warnings are currently seen.

tejlmand committed:
cmake: improve Zephyr link phase

Zephyr is a bare metal build where standard libs are disabled.

This means that c and runtime libraries must manually be linked in.

This has generally been handled by using CMake's link libraries handling
but the issue with that is both de-duplication but also library link
order.

Standard libraries must be linked at last location to ensure symbols
are always available, however this is not optimal with
target_link_libraries() because this would ultimately require every
library to know the c library to link with, which is not desired.

Therefore, setup standard C and runtime library linking in linker
CMake files for toolchains where this is required.

This commit expands the principle introduced with toolchain abstraction,
see PR#24851.

This means that a toolchain implementation may specify standard C,
runtime, C++, etc libraries, as well as their link order.
Because a property approach is used, then Zephyr modules, such as the
Picolibc module can adjust such properties.

An optional `zephyr_linker_finalize()` macro is called at the end of
Zephyr's CMakeList process and can be used by the toolchain
implementation to define the final linker invocation.

This aligns the linker handling flow to the principle introduced in
PR#24851 and improves the flexibility and robustness of Zephyr build
system.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: move specs compiler and linker flags to toolchain properties

Moving specs argument to compiler and linker properties so that the
compiler and linker in use can decide how the flags are mapped / handled
for the compiler and linker in use.

This avoids specifying `--specs=spec.picolibc` for clang which prints a
warning about an unused argument.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

jonathonpenix committed:
cmake: linker: Use the same linker for cmake checks and final build

Currently, the linker that is used when performing various cmake checks
(check_c_compiler_flag, for example) may be different than the linker that
will be used during the actual build. This happens as we currently specify
'-fuse-ld' to force the appropriate linker a) after many such checks have
already happened and b) in a way which is not automatically propagated to
check_c_compiler_flag (and friends). As a result, the toolchain's default
linker will generally be used for such checks regardless of which linker
was selected in Zephyr.

This can lead to a number of surprises when building Zephyr, particularly
when building with clang. For example:

- If the linker is misconfigured, where the build will fail can vary
  depending on whether the linker is the toolchain's default. When the
  configured linker happens to be the toolchain's default, the build
  (helpfully) fails quickly on the checks for a basic working toochain.
  When the configured linker isn't the default, the build won't fail until
  the final link steps.
- The build can fail due to issues with a linker other than the one
  configured by the user in Zephyr. For example, LLVM toolchains without
  lld will generally fail to build Zephyr (the checks for a basic
  working toochain will fail) for targets where lld is the default in LLVM
  even if GNU ld is configured in Zephyr and would otherwise be used in the
  final build.
- Flags which are only added if check_c_compiler_flag (or similar) succeeds
  may be unexpectedly omitted during the final build if the flag is
  supported in the configured linker but is unsupported in the toolchain's
  default linker (as check_c_compiler_flag will test using the default
  one).

Note that this isn't limited to clang--even when we are building with
Zephyr's SDK and force ld.bfd, we seem to use the 'ld' variant during the
cmake checks (though this generally seems fairly harmless compared to
mixing ld/lld or other proprietary linkers).

To fix this, ensure the appropriate 'fuse-ld' is set early enough and in
such a way that the same linker will be used throughout the entire build.

Signed-off-by: Jonathon Penix <jpenix@quicinc.com>
Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: PROPERTY flag support on compile options and link libraries

Extend zephyr_link_libraries to allow an optional value together with
the `zephyr_link_libraries(PROPERTY <property> [<value>])`.

This allow setting linker property combined with a value when linking
Zephyr. The value will only be applied if the property is defined.

Extend zephyr_compile_options to support the same PROPERTY flag that
has been introduced for zephyr_link_libraries().
This remove the need for developers to write complex generator
expressions for compiler flags and thus minimizes mistakes.

The following syntax is now supported in addition to the existing
syntax: `zephyr_compile_options(PROPERTY <property> [<value>])`

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: make TARGET optional in set_linker_property()

The `check_set_linker_property()` and `set_linker_property()` takes a
target argument. Make the target argument optional and use the target
`linker` as default target.

The function name `set_linker_property()` already implies that we are
setting a property and the linker target.

Remove the need to specify `TARGET linker` when using the default linker
property target.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: detect LLVM picolibc and newlib support

Newlib or Picolibc libraries for LLVM may be compiled or installed from
pre-built sources independently of LLVM itself.

This means that always indicating that TOOLCHAIN_HAS_NEWLIB=OFF and
TOOLCHAIN_HAS_PICOLIBC=OFF are wrong. But it could be just as wrong to
always indicate suport for newlib or picolibc.

Some pre-built LLVM toolchains are provided with default picolibc
support, such as LLVM for Arm embedded, but can also be used with newlib
be installing newlib add-on package.

Unfortunately it's not possible to query LLVM regarding newlib or
picolibc support.

Developers have the option of `-DTOOLCHAIN_HAS_<NEWLIB|PICOLIBC>=ON`,
but this is not widely known and cumbersome to do for each build.

An indication of newlib or picolibc support is the presence of library
specific headers, so to improve current situation we check for library
specific headers, and if those are present we assume support for the
library.

This commit improves the current support for LLVM in Zephyr when
cross-compiling, especially for users of LLVM for Arm embedded.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
manifest: update picolibc to support the new c_library property

This commit updates picolibc module to remove the need for hard-coding
linking with `-lgcc`.

Instead it sets the c_library linker property and thereby allows the
Zephyr toolchain infrastructure to properly handle the linking of
C and runtime libraries.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented on west.yml:
```diff
@@ -315,7 +315,8 @@ manifest:
       path: modules/lib/picolibc
-      revision: 764ef4e401a8f4c6a86ab723533841f072885a5b
+      revision: b9fae840f916881cd0e36b6d9883300a3da7a52a
+#      revision: pull/<no>/head
```
Fixed.

tejlmand commented:
@evgeniy-paltsev / @abrodkin can you please verify the changes with arc mwdt ?

keith-packard commented:
It doesn't look like this solves the problem of placing libc/minimal and libc/common inside the whole-archive set of libs (as in #65434)? Should we try to make that work in this series or do you think that can be done after this lands?

evgeniy-paltsev commented:
@tejlmand 

> evgeniy-paltsev / abrodkin can you please verify the changes with arc mwdt ?

Sure, I've started the verification so, I'll update about the results when they are ready. Thanks for the mentioning!

go2sh commented:
I has the same problem with the tricore llvm toolchain. Will test this pr.

tejlmand commented:
> It doesn't look like this solves the problem of placing libc/minimal and libc/common inside the whole-archive set of libs (as in https://github.com/zephyrproject-rtos/zephyr/pull/65434)? Should we try to make that work in this series or do you think that can be done after this lands?

@keith-packard thanks for directing my attention to that PR.
You're correct that this PR is not trying to solve the whole-archive problem because my focus so far has been addressing some of the older and more fundamental library link issues we have had.

But I think that the strategy used in this PR will be beneficial for #65434, so let me take a look and see if I can cover #65434 or as minimum facilitate the work you have done in #65434

keith-packard commented:
> But I think that the strategy used in this PR will be beneficial for #65434, so let me take a look and see if I can cover #65434 or as minimum facilitate the work you have done in #65434

Thanks a bunch. Getting to the point where we have explicit dependencies all the way to the C library and C runtime support and getting away from using --whole-archive for support code would really clean a bunch of stuff up, allowing us to eliminate a bunch of options which enable random functions in the common C library code (do we really need to avoid building realloc?) and saving memory on targets that don't use that functionality.

#65434 has some kludges which I'm hoping someone with way more Cmake expertise can help resolve; it's really fragile and keeps breaking as Zephyr changes underneath.

tejlmand commented:
> I has the same problem with the tricore llvm toolchain. Will test this pr.

Looking forward to hear the results.

evgeniy-paltsev commented:
@tejlmand 

I've tested it with MWDT toolchain - everything is OK, thanks!

go2sh commented:
Hey, I tested as well with HighTec Tricore Toolchain based on llvm.

Two remarks:
 * I still need the use LLD config switch to use lld as linker
 * If no LLVM_TOOLCHAIN_PATH is specified, cmake freezes due to the glob on the newlib/libc header file.

tejlmand commented:
> Hey, I tested as well with HighTec Tricore Toolchain based on llvm.

Great, thanks a lot @go2sh  :+1: 

>  I still need the use LLD config switch to use lld as linker

Correct.
LLVM supports using both ld and lld as linker, and although lld can act as a drop-in replacement then lld is still not ld, and there are behavioral differences between the two.
Therefore this PR does not address the question regarding which linker should be the default linker.

I would take changing from ld to lld as default linker as a following up PR.
Note, selecting lld over ld in Kconfig is also sub-optimal, especially regarding multi image builds (example: sysbuild), so this is an area which is also a candidate to re-think for follow-up updates.

> If no LLVM_TOOLCHAIN_PATH is specified, cmake freezes due to the glob on the newlib/libc header file.

Thanks, will try to reproduce.
This surely sounds like something needing a fix.

go2sh commented:
Thanks for the feedback on the linker, the hightec toolchain has no ld linker so, I need to add the ldd config all the time.
Would be greate to have a more general approach.

tejlmand commented:
> > If no LLVM_TOOLCHAIN_PATH is specified, cmake freezes due to the glob on the newlib/libc header file.
> 
> Thanks, will try to reproduce. This surely sounds like something needing a fix.

Fixed.

> Thanks for the feedback on the linker, the hightec toolchain has no ld linker so, I need to add the ldd config all the time.
> Would be greate to have a more general approach.

yep, I know the feeling. The same problem with the LLVM for Arm.
So there will come an improvement regarding this area in later PR.

tejlmand commented:
@keith-packard , @stephanosio please review.

keith-packard commented on cmake/linker/armlink/target.cmake:
```diff
@@ -108,5 +108,23 @@ function(toolchain_ld_link_elf)
 
+# This function will generate the correct CMAKE_C_LINK_EXECUTABLE / CMAKE_CXX_LINK_EXECUTABLE
+# rule to ensure that standard c and runtime libraries are correctly placed
+# and the end of link invocation and doesn't appear in the middle of the link
```
Is this explicit ordering working around some gaps in the dependency graph? Would cmake automatically move those libraries to the correct location if the graph was complete?

tejlmand replied:
> Is this explicit ordering working around some gaps in the dependency graph?

yes, what we are ensuring here is that we link the std libs at the exact location where they are required and not being relocated by CMake's internal logic.

This is what you usually do in a CMake toolchain file using `CMAKE_<lang>_LINK_EXECUTABLE` when building for bare-metal.

At the same time the use of properties on the linker target follows the design of compiler and linker flags and allow toolchains a uniform way of describing how standard libraries are selected and in which order (even allowing multiple times without de-duplication).

The toolchain can describe a default set, while still allowing Zephyr and Zephyr module to adjust this set (and link order) if there is a need, for example based on Kconfig settings or static libs.
Examples are:
- newlib Toolchain built-in
- picolibc built as module.

Followup to this work will be some proper documentation regarding Zephyr's toolchain infrastructure so that you should be able to create oot toolchain configurations, and even upstream those.

> Would cmake automatically move those libraries to the correct location if the graph was complete?

The problem is that this graph will never be complete.
As you have experienced, then it would require every static library created to know that it wants a specific link, like
`target_link_libraries(foo PUBLIC gcc)`

but that is not scalable and actually makes it harder to create static libs.
Technically we could wrap this for Zephyr libs, but what about libs created by regular `add_library()` CMake calls.
Also such solution would make it much harder to integrate new toolchains in future.

From design perspective, then such library link information is not something a library should ever care about.
This is also a reason why you'd normally place this in a CMake Toolchain file.
CMake toolchain files are fine for well-known project / a project with well-defined boundries but really not flexible enough for a project like Zephyr which tries to span so many use-cases and where things are flexible and can be configured through Kconfig.

keith-packard replied:
I think I managed to construct a complete graph in #65434, but the results were not terribly robust. In any case, I think this change is a good step forward.

keith-packard commented on cmake/compiler/gcc/target.cmake:
```diff
@@ -110,7 +110,6 @@ get_filename_component(LIBGCC_DIR ${LIBGCC_FILE_NAME} DIRECTORY)
 assert_exists(LIBGCC_DIR)
 
 LIST(APPEND LIB_INCLUDE_DIR "-L\"${LIBGCC_DIR}\"")
```
It seems like these two values need to go together? Why is LIBGCC_DIR being handled here while libgcc itself is moved elsewhere?

tejlmand replied:
thanks for spotting.
It probably should, especially as the usage of `LIB_INCLUDE_DIR` was adjusted for clang so will take a look.

As for the `gcc` move, then that's moved because any value which is generally pre-known by the toolchain (`-lgcc`) and therefore should be present in a `linker_libraries.cmake` file, but the actual path is not known at this time.
The path is fetched by querying the compiler, and that cannot be done until the target arch (and thus the correct target compiler) has been decided.

That said, generally it should not be necessary to first query the compiler about `--print-libgcc-file-name` just to feed this information back to the linker later. Specifically for Zephyr SDK this is in principle not needed.

Where customizing `-L` / `LIB_INCLUDE_DIR` makes most sense is when SYSROOT_DIR is used to direct the linker to specific library location, as well as the `cross-compile` toolchain usage.
The core of the current implementation originates back to the KBuild transition days, https://github.com/zephyrproject-rtos/zephyr/commit/12f8f7616567f224b920512b6f81a65421e30bae.
However, revisiting current implementation is outside scope of this PR.

keith-packard replied:
That all sounds reasonable; feel free to leave this as-is for now; I'll be re-working #65434 once this has landed (and I have enough spare time).

tejlmand replied:
aligned to use linker properties.

keith-packard commented on cmake/linker/ld/linker_libraries.cmake:
```diff
@@ -0,0 +1,33 @@
+# Do not specify default link libraries when targeting host (native build).
+if(NOT CONFIG_NATIVE_BUILD)
+  set_linker_property(NO_CREATE TARGET linker PROPERTY c_library    "-lc")
+  set_linker_property(NO_CREATE TARGET linker PROPERTY rt_library   "-lgcc")
```
Yeah, it would be nice to see the libgcc path defined here so that it was tied with the library name. I haven't dug into how that value is computed though.

tejlmand replied:
it's probably nice to have it with a comment regarding why it's empty at this location and where it is being filled.

tejlmand replied:
done.

tejlmand commented:
fyi @jonathonpenix

keith-packard reviewed (dismissed):
Looks like there are still some test regressions that need fixing, but overall this is looking good.

stephanosio reviewed (dismissed):
Definitely looks to be an improvement.

tejlmand commented:
> Looks like there are still some test regressions that need fixing, but overall this is looking good

actually seems to be because https://github.com/zephyrproject-rtos/zephyr/pull/78306 has been introduced, which doesn't like the changes in this PR. Still investigating.

tejlmand committed:
manifest: update zephyr-lang-rust to always link runtime libs first

Update zephyr-lang-rust module to include commit for linking the
runtime library before the librustapp.a.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

carlescufi commented:
@keith-packard @stephanosio could you take another look?

stephanosio commented:
Looks ok to me. https://github.com/zephyrproject-rtos/zephyr-lang-rust/pull/9 needs to be merged before this can be merged.

carlescufi commented:
> Looks ok to me. [zephyrproject-rtos/zephyr-lang-rust#9](https://github.com/zephyrproject-rtos/zephyr-lang-rust/pull/9) needs to be merged before this can be merged.

Merged @tejlmand

keith-packard reviewed (dismissed):
The picolibc changes look reasonable to me.

keith-packard commented on cmake/linker/ld/linker_libraries.cmake:
```diff
@@ -0,0 +1,38 @@
+  # This is needed because when linking with newlib on aarch64, then libgcc has a
+  # link dependency to libc (strchr), but libc also has dependencies to libgcc.
+  # Lib C depends on libgcc. e.g. libc.a(lib_a-fvwrite.o) references __aeabi_idiv
+  set_property(TARGET linker APPEND PROPERTY link_order_library "math;c;rt;c")
```
Would it be reasonable to use `--start-group -lc -lgcc --end-group` instead of listing libc twice?

tejlmand replied:
Personally I think that it's good to make it visible that there are indeed circular dependencies, and that we can see the number of loops.
`--start/end-group` are useful when you really start to have complex circular dependencies, but it easily turn into cargo-cult which no one dares to cleanup, and in future easily ends in a bunch of libs inside those flags.

As side-note, from the ld manual:
> `--start-group archives --end-group`
>    ...
>    Using this option has a significant performance cost. It is best to use it only when there are unavoidable circular references between two or more archives.

keith-packard commented on cmake/linker/lld/linker_libraries.cmake:
```diff
@@ -0,0 +1,21 @@
+set_linker_property(NO_CREATE TARGET linker PROPERTY c_library "-lc")
+# Default per standard, will be populated by clang/target.cmake based on clang output.
+set_linker_property(NO_CREATE TARGET linker PROPERTY rt_library "")
+set_linker_property(TARGET linker PROPERTY c++_library "-lc++;-lc++abi")
```
should this mention `math_library` somehow? Either noting that it's not needed or setting to the empty string?

tejlmand replied:
II don't think so.
The lib names are not fixed in that sense, a linker implementation can define extra libraries to always link in, as described here:
https://github.com/zephyrproject-rtos/zephyr/blob/3b8e621e9cd6b8382ccdaad4bd757012dafe3d59/cmake/linker/linker_libraries_template.cmake#L9-L10

I will be working on a docset to describe the whole toolchain infrastructure and how to implement support for a custom one.
And there the math lib could be a good candidate to show.

keith-packard commented on cmake/compiler/compiler_flags_template.cmake:
```diff
@@ -139,3 +139,7 @@ set_compiler_property(PROPERTY warning_shadow_variables)
 set_compiler_property(PROPERTY no_builtin)
 set_compiler_property(PROPERTY no_builtin_malloc)
+
+# Compiler flag for defining specs. Used only by gcc, other compilers may keep
```
Should we eventually plan on supporting clang's `--config` using this same mechanism?

tejlmand replied:
I was thinking of that, and some clang/LLVM releases, like LLVM for Arm, does provide dedicated configs for this, but i'm unsure how well established / standardized this principle is, so for a start it's kept a bit independent in `cmake/compiler/clang/target.cmake`.
and also because it's needed for the existing compile check functionality.

keith-packard commented on cmake/compiler/clang/target.cmake:
```diff
@@ -64,6 +64,42 @@ if(NOT "${ARCH}" STREQUAL "posix")
+  endif()
+
+  if(CONFIG_PICOLIBC)
+    file(GLOB_RECURSE picolibc_cfg ${LLVM_TOOLCHAIN_PATH}/picolibc.cfg)
```
maybe share the --specs mechanism somehow?

tejlmand replied:
see other answer related to this. But good to have in mind for further improvements :+1:

nashif commented:
@tejlmand  This change is breaking building with cadence xtensa toolchains:

```

): in function `intel_adsp_ipc_init':
/home/to/zephyrproject/zephyr/soc/intel/intel_adsp/common/ipc.c:101: undefined reference to `memset'
/home/nashif/xtensa/XtDevTools/install/tools/RI-2022.10-linux/XtensaTools/bin/xtensa-elf-ld: zephyr/kernel/libkernel.a(thread.c.obj):(.literal.z_impl_k_thread_name_set+0x0): undefined reference to `strncpy'
/home/nashif/xtensa/XtDevTools/install/tools/RI-2022.10-linux/XtensaTools/bin/xtensa-elf-ld: zephyr/kernel/libkernel.a(thread.c.obj): in function `z_impl_k_thread_name_set':
/home/to/zephyrproject/zephyr/kernel/thread.c:145: undefined reference to `strncpy'
/home/nashif/xtensa/XtDevTools/install/tools/RI-2022.10-linux/XtensaTools/bin/xtensa-elf-ld: zephyr/kernel/libkernel.a(thread.c.obj): in function `z_setup_new_thread':
/home/to/zephyrproject/zephyr/kernel/thread.c:610: undefined reference to `strncpy'
/home/nashif/xtensa/XtDevTools/install/tools/RI-2022.10-linux/XtensaTools/bin/xtensa-elf-ld: zephyr/kernel/libkernel.a(pipes.c.obj): in function `pipe_xfer':
/home/to/zephyrproject/zephyr/kernel/pipes.c:205: undefined reference to `memcpy'
/home/nashif/xtensa/XtDevTools/install/tools/RI-2022.10-linux/XtensaTools/bin/xtensa-elf-ld: /home/to/zephyrproject/zephyr/kernel/pipes.c:212: undefined reference to `memcpy'
clang-10: error: Xtensa-ld command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /home/nashif/.local/bin/cmake --build /home/to/zephyrproject/zephyr/build --target run

```

marc-hb commented on cmake/linker/ld/target.cmake:
```diff
@@ -137,14 +126,44 @@ function(toolchain_ld_link_elf)
     $<TARGET_OBJECTS:${OFFSETS_LIB}>
-    ${LIB_INCLUDE_DIR}
     -L${PROJECT_BINARY_DIR}
-    ${TOOLCHAIN_LIBS}
```
Now `-lhal` is missing when using `xt-xcc` and compilation fails with undefined `xthal_window_spill` (`xthal_window_spill` is provided by the `xt-xcc` toolchain in `libhal.a`)

If I restore this specific line then `xt-xcc` compilation succeeds again.

I wonder why was this `${TOOLCHAIN_LIBS}` line removed from SOME `target.cmake` files by commit 2e3873adde2 but as of today's commit 27456ed2b57f it is still present in the following files? This seems inconsistent.

```
cmake/compiler/icx/target.cmake
cmake/linker/arcmwdt/target.cmake
cmake/linker/xt-ld/target.cmake
```

Indeed, if I remove `${TOOLCHAIN_LIBS}` from `cmake/linker/xt-ld/target.cmake` then `xt-clang` fails with the same error.

cc: @kv2019i , @andyross, @cujomalainey, @dcpleung 

FYI:

```
cmake/compiler/xcc/target.cmake:list(APPEND TOOLCHAIN_LIBS
cmake/compiler/xcc/target.cmake-  gcc
cmake/compiler/xcc/target.cmake-  hal
cmake/compiler/xcc/target.cmake-  )
```

tejlmand replied:
an oversight.

I initially wanted to still allow for the old use of `TOOLCHAIN_LIBS` for toolchains I didn't have available, but later changes related to picolibc has actually resulted in the situation that all toolchains must be updated.

marc-hb replied:
Thanks @tejlmand !

> actually resulted in the situation that all toolchains must be updated.

While @dcpleung is removing the obsolete TOOLCHAIN_LIBS for Xtensa in #79731, as of today I see it still present in the main branch for `cmake/compiler/icx/target.cmake` and `cmake/linker/arcmwdt/target.cmake`. I think it would be good to file a bug and use `git blame` to warn the owners of these files that TOOLCHAIN_LIBS does not work anymore in case they still try and fail to use it downstream somehow. Something I did not mention before: it took me HOURS to debug this Xtensa compilation failure and find that it was down to this TOOLCHAIN_LIBS variable. Funny enough, `git grep hal` does not work very well! It would be nice to avoid others' a similar... effort.

In case these toolchains have no owner left, then it is a great opportunity to notice that and delete them entirely.

There is also a mysterious `TOOLCHAIN_LIBS_OBJECTS` in `cmake/linker/armlink/target.cmake`... My 2 cents.
